juju/terraform-provider-juju
# Add support for Placement directives in applications and add a new Machine resource

## Description

This PR implements two parts of the Juju's functionality. The first part is defining a placement directive in Charm applications to instruct Juju where to deploy Charm units (this is the equivalent of the `--to` Juju CLI argument). The second part adds a new Machine resource that allows you to use Juju to create machines and manage them through Terraform (this is meant to do the same thing as the `add-machine` command in Juju).

I wrote this extension to the Terraform Juju provider because I wanted to use the Terraform Juju provider to deploy Charmed OpenStack as outlined in the [official upstream Charmed OpenStack deployment guide](https://docs.openstack.org/project-deploy-guide/charm-deployment-guide/latest/install-openstack.html). I could not do that without placement directives because Juju would create a new machine for every application. My aim is to deploy OpenStack on as few machines as possible for efficiency. I added the placement directive to the Application resource schema and Application client alongside the existing arguments that are passed to the Juju client.

However, I realized that the placement directive requires that you know the machine IDs ahead of time. I was not going to know the machine IDs ahead of time because I wanted to dynamically create and destroy the machines I use to run OpenStack. I chose to implement a basic Machine resource in the Terraform Juju provider so that I can create the manages and manage them in Terraform alongside the applications. With a Machine resource, I can dynamically get the Juju machine ID and pass that into the Application resource definitions.

The code I added to the Terraform Juju provider works. I can use Terraform to deploy Charmed OpenStack on four machines with storage, compute, and control plane services on the same machines (that is, a hyperconverged deployment). I have [a Terraform plan in a rough draft state](https://github.com/jadonn/terraform-juju-openstack-testing/blob/main/main.tf) in a personal GitHub repo that will deploy everything for OpenStack using MAAS.

The overall goal for this work is to orchestrate automating the deployment of OpenStack so I can set up frequent automated testing of OpenStack. I really like the Terraform Juju provider for this project because I do not have to write a great deal of custom code that would only ever be used for this one project. Using the Terraform Juju provider significantly reduces my overhead and simplifies the configuration of my deployment considerably.

## Type of change

Please mark if proceeds.


## Environment

- Juju controller version: 2.9
- Terraform version: 1.3

## QA steps

You should be able to test this PR with any Juju cloud. During development, I used a LXD cloud running on my local machine to test provisioning machines and using the placement directive. You should be able to run the following simple plan to verify the core functionality of creating a machine and assigning units with the placement directive

```tf
provider juju {}

resource "juju_model" "machine_test" {
  name = "machine-test"
}

resource "juju_machine" "this" {
  depends_on = [juju_model.machine_test]
  model = juju_model.machine_test.name
  series = "focal"
  name = "testMachine"
}

resource "juju_machine" "that" {
  model = juju_model.machine_test.name
  series = "focal"
  name = "thatMachine"
}

resource "juju_application" "hello" {
  model = juju_model.machine_test.name
  charm {
    name = "hello-juju"
  }
  placement = join(",", [split(":", juju_machine.this.id)[1], split(":", juju_machine.that.id)[1]])
  units = 2
}
```

This plan should create two machines using Ubuntu Focal and deploy one unit of the `hello-juju` application to each machine. After the plan runs, you should see the correct machines and placement when you use `juju status` to see the model.

# Additional notes

Let me know what changes need to be made or if you would like to take a different approach to representing machines or placement directives. I am happy to make changes to follow your overall plans or goals better.
